cpp/ql/src/semmle/code/cpp/exprs/Call.qll

```diff
@@ -134,6 +153,110 @@ abstract class Call extends Expr, NameQualifiableElement {
   }
 }
 
+private newtype TArgumentID =
```

Comment:
I haven't been in the loop about why this type is so important for this PR. Why wouldn't it suffice to write

```
class ArgumentID extends int {
  ArgumentID() { this = [ -1 .. max(any(Call c).getNumberOfArguments()) - 1 ] }
}
```

That might cause some code elsewhere to become a bit more verbose, but this commit adds 135 lines of QL. That's very verbose.

Reply:
In this PR, it will make it possible to give useful labels to side effect nodes without associated parameters - see https://github.com/Semmle/ql/pull/1735#discussion_r313856003

Unrelated to this PR, I think it will make function modeling easier going forward, but we could put that off.

Reply:
That `ArgumentID` class is not what I was hoping for when I asked https://github.com/Semmle/ql/pull/1735#discussion_r313856003. If the purpose is to print something useful in the debug output, I'd be happy with just printing the argument index as an `int` for all types of argument (the qualifier can be `-1`).

Remember that `FunctionCall.getTarget` can have multiple results when linker awareness is incomplete, so there's always a risk of printing something confusing if you try to name arguments based on matching them to parameters. One result might use varargs, and the other might not. Our implementation of `Parameter.getName` is also full of heuristics, which is another reason why I don't think it's good for debug output.

And then there's `ExprCall`, where I don't believe there are `Parameter`s or names available at the call site. I'm not even sure we extract whether a `RoutineType` has varargs.